SwiftCompilerSources/Sources/Optimizer/Utilities/EscapeInfoWalker.swift

```diff
@@ -0,0 +1,834 @@
+      is InitExistentialMetatypeInst, is OpenExistentialMetatypeInst,
+      is ExistentialMetatypeInst, is DeallocRefInst, is SetDeallocatingInst,
+      is FixLifetimeInst, is ClassifyBridgeObjectInst, is BridgeObjectToWordInst,
+      is EndBorrowInst, is EndAccessInst,
+      is StrongRetainInst, is RetainValueInst,
+      is ClassMethodInst, is SuperMethodInst, is ObjCMethodInst,
+      is ObjCSuperMethodInst, is WitnessMethodInst,is DeallocStackRefInst:
+      return false
+    default:
+      return isEscaping
+    }
+    return false
+  }
+  
+  mutating func leafUse(value: Operand, path: Path, state: State) -> Bool {
+    return leafUse(any: value, path: path, state: state)
+  }
+  
+  mutating func leafUse(address: Operand, path: Path, state: State) -> Bool {
+    return leafUse(any: address, path: path, state: state)
+  }
+  
+  mutating func shouldRecomputeDown(def: Value, path: Path, state: State) -> (Path, State)? {
+    if let entry = walkedDownCache[def.hashable, default: CacheEntry()].needWalk(path: path, followStores: state.followStores, knownType: state.knownType) {
+      return (entry.path, state.with(knownType: entry.knownType).with(followStores: entry.followStores))
+    }
+    return nil
+  }
+  
+  mutating func cachedWalkDown(def: Value, path: Path, state: State) -> Bool {
+    if let (path, state) = shouldRecomputeDown(def: def, path: path, state: state) {
+      if def.type.isAddress {
+        return walkDownUses(address: def, path: path, state: state)
+      } else {
+        return walkDownUses(value: def, path: path, state: state)
+      }
+    } else {
+      return false
+    }
+  }
+  
+  mutating func walkUp(value: Value, path: Path, state: State) -> Bool {
+    if hasRelevantType(value, at: path) {
+      if DEBUGWALK {
+        print("visitDef \"\(path)\": \(value)")
+      }
+      switch visitor.visitDef(def: value, path: path, state: state) {
+      case .continueWalkUp:
+        return walkUpDefault(value: value, path: path, state: state)
+      case .walkDown:
+        return cachedWalkDown(def: value, path: path, state: state.with(knownType: nil))
+      case .ignore:
+        return false
+      case .abort:
+        return true
+      }
+    }
+    return false
+  }
+  
+  mutating func walkUp(address: Value, path: Path, state: State) -> Bool {
+    if hasRelevantType(address, at: path) {
+      if DEBUGWALK {
+        print("visitDef \"\(path)\": \(address)")
+      }
+      switch visitor.visitDef(def: address, path: path, state: state) {
+      case .continueWalkUp:
+        return walkUpDefault(address: address, path: path, state: state)
+      case .walkDown:
+        return cachedWalkDown(def: address, path: path, state: state)
+      case .ignore:
+        return false
+      case .abort:
+        return true
+      }
+    }
+    return false
+  }
+  
+  mutating func rootDef(any def: Value, path: SmallProjectionPath, state: State) -> Bool {
```

Comment:
The same here: Why not handling the value/address instructions directly in the dedicated `rootDef(value:)`/`rootDef(address:)` functions?